File: common/src/main/java/com/vladte/devhack/common/service/domain/global/VacancyService.java
package com.vladte.devhack.common.service.domain.global;

import com.vladte.devhack.common.service.domain.CrudService;
import com.vladte.devhack.entities.Vacancy;

import java.util.List;
import java.util.UUID;

/**
 * Service interface for managing Vacancy entities.
 */
public interface VacancyService extends CrudService<Vacancy, UUID> {

    /**
     * Find vacancies by company name.
     *
     * @param companyName The company name to search for
     * @return A list of vacancies from the specified company
     */
    List<Vacancy> findByCompanyName(String companyName);

    /**
     * Find vacancies by position.
     *
     * @param position The position to search for
     * @return A list of vacancies with the specified position
     */
    List<Vacancy> findByPosition(String position);

    /**
     * Find vacancies by source.
     *
     * @param source The source to search for
     * @return A list of vacancies from the specified source
     */
    List<Vacancy> findBySource(String source);

    /**
     * Find vacancies by remote allowed status.
     *
     * @param remoteAllowed The remote allowed status to search for
     * @return A list of vacancies with the specified remote allowed status
     */
    List<Vacancy> findByRemoteAllowed(Boolean remoteAllowed);

    /**
     * Search vacancies by keyword in company name, position, or technologies.
     *
     * @param keyword The keyword to search for
     * @return A list of vacancies matching the keyword
     */
    List<Vacancy> searchByKeyword(String keyword);
}

File: angular-ui/src/app/services/global/vacancy.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {VacancyDTO} from '../../models/global/vacancy.model';
import {BaseService} from '../base.service';

/**
 * Service for managing vacancies.
 * Extends BaseService to inherit common CRUD operations and follows OOP principles.
 */
@Injectable({
    providedIn: 'root'
})
export class VacancyService extends BaseService<VacancyDTO> {
    protected baseUrl = '/api/vacancies';

    constructor(http: HttpClient) {
        super(http);
    }

    /**
     * Search vacancies with multiple criteria
     * @param query - Search query
     * @param location - Location filter
     * @param company - Company filter
     * @returns Observable array of vacancies
     */
    override search(searchParams: { [key: string]: string | undefined }): Observable<VacancyDTO[]>;
    override search(keyword?: string, location?: string, company?: string): Observable<VacancyDTO[]>;
    override search(queryOrParams?: string | {
        [key: string]: string | undefined
    }, location?: string, company?: string): Observable<VacancyDTO[]> {
        if (typeof queryOrParams === 'object') {
            // Ensure we map common aliases to backend-expected params
            const params: { [key: string]: string | undefined } = {...queryOrParams};
            if (params['query'] && !params['keyword']) {
                params['keyword'] = params['query'];
                delete params['query'];
            }
            return super.search(params);
        }
        // Backend expects 'keyword' param for search
        return super.search({keyword: queryOrParams, location, company});
    }

    /**
     * Get vacancies by company
     * @param company - Company name to filter by
     * @returns Observable array of vacancies
     */
    getByCompany(company: string): Observable<VacancyDTO[]> {
        return this.http.get<VacancyDTO[]>(`${this.baseUrl}/by-company/${company}`);
    }

    /**
     * Get vacancies by location
     * @param location - Location to filter by
     * @returns Observable array of vacancies
     */
    getByLocation(location: string): Observable<VacancyDTO[]> {
        return this.http.get<VacancyDTO[]>(`${this.baseUrl}/by-location/${location}`);
    }

    /**
     * Get recent vacancies
     * @param limit - Optional limit for number of vacancies
     * @returns Observable array of recent vacancies
     */
    getRecent(limit?: number): Observable<VacancyDTO[]> {
        let params = new HttpParams();
        if (limit) params = params.set('limit', limit.toString());

        return this.http.get<VacancyDTO[]>(`${this.baseUrl}/recent`, {params});
    }

    /**
     * Filter by remote allowed flag using existing REST endpoint
     */
    getByRemote(remoteAllowed: boolean): Observable<VacancyDTO[]> {
        const params = new HttpParams().set('remoteAllowed', String(remoteAllowed));
        return this.http.get<VacancyDTO[]>(`${this.baseUrl}/by-remote`, {params});
    }

    /**
     * Parse vacancy from URL
     * @param url - URL to parse vacancy from
     * @returns Observable parsed vacancy
     */
    parseVacancy(url: string): Observable<VacancyDTO> {
        const body = {url};
        return this.http.post<VacancyDTO>(`${this.baseUrl}/parse`, body);
    }
}
